scripts/validate_docs_snippet_line_numbers.py

```diff
@@ -0,0 +1,202 @@
+
+
+def _collect_docusaurus_refs(
+    file: str, reference_map: Dict[str, List[DocusaurusRef]]
+) -> None:
+    with open(file) as f:
+        contents = f.read()
+
+    references: List[Tuple[str, str, str]] = DOCUSAURUS_REGEX.findall(contents)
+
+    for reference in references:
+        docusaurus_ref = DocusaurusRef.from_regex_parsed_values(
+            file=file, values=reference
+        )
+        reference_map[docusaurus_ref.py_file].append(docusaurus_ref)
+        logger.debug(
+            "Create association between %s and the docs that use it",
+            docusaurus_ref.py_file,
+        )
+
+
+def evaluate_snippet_validity(
+    docusaurus_refs: Dict[str, List[DocusaurusRef]]
+) -> List[DocusaurusRef]:
+    """Given a series of references, reviews the referenced files and ensures that
+    opening and closing tags directly surround the snippet.
+
+    Ex: ```python file=...#L2
+
+        1 <snippet>
+        2
+        3 context = DataContext()
+        4 </snippet>
+
+    Note that the tags do not need to be immediately before and after the content - as long
+    as the lines in between are whitespace, the algorithm will not flag it as an issue.
+
+    Args:
+        docusaurus_refs: A map between Python files and the docs that reference them.
+
+    Returns:
+        A list of broken references.
+    """
+    all_broken_refs: List[DocusaurusRef] = []
+    for docs_file, ref_list in docusaurus_refs.items():
+        logger.debug(
+            "Checking %s for snippet validity",
+            docs_file,
+        )
+        file_broken_refs: List[DocusaurusRef] = _evaluate_snippet_validity(
+            docs_file, ref_list
+        )
+        all_broken_refs.extend(file_broken_refs)
+
+    return all_broken_refs
+
+
+def _evaluate_snippet_validity(
+    file: str, ref_list: List[DocusaurusRef]
+) -> List[DocusaurusRef]:
+    with open(file) as f:
+        file_contents: List[str] = f.readlines()
+
+    broken_refs: List[DocusaurusRef] = []
+    for ref in ref_list:
+        if not _validate_docusaurus_ref(ref, file_contents):
+            broken_refs.append(ref)
+
+    return broken_refs
+
+
+def _validate_docusaurus_ref(
+    docusaurus_ref: DocusaurusRef, file_contents: List[str]
+) -> bool:
+    start_line, end_line = docusaurus_ref.line_numbers
+    if start_line > len(file_contents) or end_line > len(file_contents):
+        logger.warning(
+            "Invalid snippet due to out-of-bounds reference: %s", docusaurus_ref
+        )
+        return False
+
+    return _validate_opening_tag(start_line, file_contents) and _validate_closing_tag(
+        end_line, file_contents
+    )
+
+
+def _validate_opening_tag(start_line: int, file_contents: List[str]) -> bool:
+    ptr: int = start_line - 1
+
+    line: str
+    while ptr >= 0:
+        line = file_contents[ptr - 1]
+        if OPENING_TAG in line:
+            return True
+        elif not line.isspace():
+            return False
+        ptr -= 1
+
+    return False
```

Comment:
Will iterate until it hits the appropriate tag. If we hit some other text, we know that our tags are improper and the snippet is determined to be invalid. `line.isspace()` will check whether or not a given line is entirely composed of whitespace.